arangod/Agency/TransactionBuilder.h

```diff
@@ -32,210 +32,214 @@
-    _buffer.openObject();
-    return write_trx<B, envelope>(std::move(_buffer));
+  T* release() {
+    T* result = _ptr;
+    _ptr = nullptr;
+    return result;
   }
+  void reset(T* p) { _ptr = p; }
 
-  void done() && {
-    _buffer.closeArray();
-  }
-
-  static envelope create(VPackBuilder &b) {
-    buffer buff(b);
-    envelope env(buffer{b});
-    env._buffer.openArray();
-    return env;
-  }
+  T* operator->() { return _ptr; }
+  T* get() { return _ptr; }
+  T& operator*() { return *_ptr; }
+  operator bool() { return _ptr != nullptr; }
 
  private:
-  friend write_trx<B, envelope>;
-  friend read_trx<B, envelope>;
-  friend precs_trx<B, envelope>;
-  explicit envelope(buffer&& b) : _buffer(std::move(b)) {}
-  envelope() = default;
-  buffer _buffer;
+  T* _ptr = nullptr;
 };
 
-namespace detail {
-
-template<typename V>
+template <typename V>
 void add_to_builder(VPackBuilder* b, V const& v) {
   b->add(VPackValue(v));
 }
 
-template<>
+template <>
 inline void add_to_builder(VPackBuilder* b, VPackSlice const& v) {
   b->add(v);
 }
+}  // namespace detail
 
-}
-
-
-template<>
-struct buffer_mapper<VPackBuilder> {
-
-  explicit buffer_mapper(VPackBuilder& builder) : _builder(&builder) {};
-
-  template<typename K, typename V>
-  void setKey(K&& k, V&& v) {
-      setValue(std::forward<K>(k));
-      setValue(std::forward<V>(v));
+struct envelope {
+  using builder_ptr = detail::moving_ptr<VPackBuilder>;
+
+  struct read_trx {
+    envelope end() {
+      _builder->close();
+      return envelope(_builder.release());
+    }
+    template <typename K>
+    read_trx& key(K&& k) {
+      detail::add_to_builder(_builder.get(), std::forward<K>(k));
+      return *this;
+    }
+    ~read_trx() try {
+      if (_builder) {
+        end();
+      }
+    } catch (...) {
+    }
+    read_trx(read_trx&&) = default;
+    read_trx& operator=(read_trx&&) = default;
+
+   private:
+    friend envelope;
+    read_trx(builder_ptr b) : _builder(std::move(b)) { _builder->openArray(); }
+    builder_ptr _builder;
+  };
+
+  struct write_trx;
+  struct precs_trx {
+    template <typename K, typename V>
+    precs_trx&& isEqual(K&& k, V&& v) && {
+      detail::add_to_builder(_builder.get(), std::forward<K>(k));
+      _builder->openObject();
+      detail::add_to_builder(_builder.get(), "old");
+      detail::add_to_builder(_builder.get(), std::forward<V>(v));
+      _builder->close();
+      return std::move(*this);
+    }
+
+    template <typename K>
+    precs_trx&& isEmpty(K&& k) && {
+      detail::add_to_builder(_builder.get(), std::forward<K>(k));
+      _builder->openObject();
+      _builder->add("oldEmpty", VPackValue(true));
+      _builder->close();
+      return std::move(*this);
+    }
+
+    envelope end() {
+      _builder->close();
+      _builder->add(VPackValue(AgencyWriteTransaction::randomClientId()));
+      return envelope(_builder.release());
+    }
+    ~precs_trx() try {
+      if (_builder) {
+        end();
```

Comment:
agan, I think this is more work than necessary. A `_builder.release()` should be enough here, and also allow getting rid of the try...catch statements.